boilerplate/app/components/Screen.tsx

```diff
@@ -0,0 +1,243 @@
+import React, { useRef, useState } from "react"
+import {
+  KeyboardAvoidingView,
+  KeyboardAvoidingViewProps,
+  LayoutChangeEvent,
+  Platform,
+  ScrollView,
+  ScrollViewProps,
+  StatusBar,
+  StatusBarProps,
+  StyleProp,
+  View,
+  ViewStyle,
+} from "react-native"
+import { Edge, SafeAreaView, SafeAreaViewProps } from "react-native-safe-area-context"
+import { colors } from "../theme"
+
+interface BaseScreenProps {
+  /**
+   * Children components.
+   */
+  children?: React.ReactNode
+  /**
+   * Style for the outer content container useful for padding & margin.
+   */
+  style?: StyleProp<ViewStyle>
+  /**
+   * Style for the inner content container useful for padding & margin.
+   */
+  contentContainerStyle?: StyleProp<ViewStyle>
+  /**
+   * Override the default edges for the safe area.
+   */
+  safeAreaEdges?: Edge[]
+  /**
+   * Background color
+   */
+  backgroundColor?: string
+  /**
+   * Status bar setting. Defaults to dark-content.
+   */
+  statusBarStyle?: "light-content" | "dark-content"
+  /**
+   * By how much should we offset the keyboard? Defaults to 0.
+   */
+  keyboardOffset?: number
+  /**
+   * Pass any additional props directly to the SafeAreaView component.
+   */
+  SafeAreaViewProps?: SafeAreaViewProps
+  /**
+   * Pass any additional props directly to the StatusBar component.
+   */
+  StatusBarProps?: StatusBarProps
+  /**
+   * Pass any additional props directly to the KeyboardAvoidingView component.
+   */
+  KeyboardAvoidingViewProps?: KeyboardAvoidingViewProps
+}
+
+interface FixedScreenProps extends BaseScreenProps {
+  preset?: "fixed"
+}
+interface ScrollScreenProps extends BaseScreenProps {
+  preset?: "scroll"
+  /**
+   * Should keyboard persist on screen tap. Defaults to handled.
+   * Only applies to scroll preset.
+   */
+  keyboardShouldPersistTaps?: "handled" | "always" | "never"
+  /**
+   * Pass any additional props directly to the ScrollView component.
+   */
+  ScrollViewProps?: ScrollViewProps
+}
+
+interface AutoScreenProps extends Omit<ScrollScreenProps, "preset"> {
+  preset?: "auto"
+  /**
+   * Threshold to trigger the automatic disabling/enabling of scroll ability.
+   * Defaults to `{ percent: 0.92 }`.
+   */
+  scrollEnabledToggleThreshold?: { percent?: number; point?: number }
+}
+
+export type ScreenProps = ScrollScreenProps | FixedScreenProps | AutoScreenProps
```

Comment:
Makes the various screen types a discriminating union using the `preset` prop. This allows for better vscode suggestions for the prop differences.